src/tiled/objectselectiontool.cpp

```diff
@@ -595,24 +593,25 @@ void ObjectSelectionTool::mouseReleased(QGraphicsSceneMouseEvent *event)
             auto underlyingObjects = objectItemsAt(event->scenePos());
             if (underlyingObjects.isEmpty())
                 break;
-            int currRotationIndex = underlyingObjects.indexOf(mSelectedInUnderlyingRotation);
-            mClickedObjectItem = mSelectedInUnderlyingRotation
-                    = underlyingObjects.at((currRotationIndex + 1) % underlyingObjects.size());
+
+            int lastSelectedIndex = -1;
+            for (auto selected : selection)
+                lastSelectedIndex = std::max(lastSelectedIndex, underlyingObjects.indexOf(selected));
+            do lastSelectedIndex = (lastSelectedIndex + 1) % underlyingObjects.size();
+            while (selection.contains(underlyingObjects.at(lastSelectedIndex))
+                   && lastSelectedIndex != underlyingObjects.size() - 1);
+            mClickedObjectItem = underlyingObjects.at(lastSelectedIndex);
         }
         if (mClickedObjectItem) {
             if (modifiers & (Qt::ShiftModifier | Qt::ControlModifier)) {
-                if (selection.contains(mClickedObjectItem))
-                    selection.remove(mClickedObjectItem);
+                if (!(modifiers & Qt::AltModifier) && selection.contains(mClickedObjectItem))
+                    selection.remove(mClickedObjectItem);// Removal is not supported in alt+click mode
```

Comment:
Otherwise it will select and unselect the same object in the cycle, I think it looks buggy